### PR TITLE
Added max dateAdded to artistview and albumview

### PR DESCRIPTION
**Pull request by razzeee**
- Also removed the unused fields on album and artist
- Make them sortable and show them

@Montellese

### CODE REVIEW

File: xbmc/SortFileItem.h [hunk 1]
```diff
@@ -40,6 +40,8 @@ typedef enum {
   SORT_METHOD_YEAR,
   SORT_METHOD_VIDEO_RATING,
   SORT_METHOD_DATEADDED,
+  SORT_METHOD_ALBUM_DATEADDED,
+  SORT_METHOD_ARTIST_DATEADDED,
```
Comment (MartijnKaijser) [inline xbmc/SortFileItem.h]: this will break the int we use in plugins http://mirrors.xbmc.org/docs/python-docs/14.x-helix/xbmcplugin.html#-addSortMethod

Comment (ronie) [inline xbmc/SortFileItem.h]: perhaps we should update the python docs and recommend to use a string instead of an int?
95% of our plugins already use that.

http://xbmclogs.com/ptjsjxwyn/vlcvps/raw

Comment (MartijnKaijser) [inline xbmc/SortFileItem.h]: right, if that already working (never really looked into it) then it's OK for me.
Let's update the pydocs properly with a sorted list

Comment (Razzeee) [inline xbmc/SortFileItem.h]: Anyway, I will move these two declarations to the end of the enum

Comment (MartijnKaijser) [inline xbmc/SortFileItem.h]: neah keep them there. plugins will just need to be updated

Comment (Razzeee) [inline xbmc/SortFileItem.h]: Okay, can do ;)

Comment (ronie) [inline xbmc/SortFileItem.h]: /me suggests an python api bump then.

**Comment by Tolriq** [comment]
Please expose this to json (with song dateadded) 

It's just a matter of updating https://github.com/xbmc/xbmc/tree/master/xbmc/interfaces/json-rpc/schema and would simplify smart syncing for remotes to know the date of the last added items to trigger syncs.

**Comment by Razzeee** [comment]
@Tolriq 
good thinking! Will do.

**Comment by Montellese** [comment]
Why did you add new properties to `CFileItem` instead of simply reusing the existing dateadded property? A `CFileItem` can either represent an artist, album or song and I don't think there's any use case where the user wants to sort all albums by the artist's dateadded.

**Comment by Razzeee** [comment]
@Montellese 
you mean ```CAlbum``` and ```CArtist```?
Wasn't aware of the existing, tried to use the one on the tag, but I need to get the data from somewhere.
I'm also not sure about the usecase, but thought that might be what you asked for in the other PR.

**Comment by Montellese** [comment]
@Razzeee: Adding them to `CAlbum` and `CArtist` is fine. I was talking about the `m_albumDateAdded` and `m_artistDateAdded` which you added to `CMusicInfoTag` (sorry not `CFileItem`). IMO that doesn't make any sense. You should just use the existing `m_dateAdded` field which will also save you from having to add new sort and format label implementations. It will also require less changes to JSON-RPC because you only have to add the `"dateadded"` property to the artist and album properties.

BTW what do you mean by "I can't figure out why the Artist Class fails on me when I comment this code in.". What exactly does "fails on me" mean? Compiler error, runtime error, ...?

**Comment by Razzeee** [comment]
Compile error. Something along the lines of  CDateTime artistDateAdded is not declared.

**Comment by Montellese** [comment]
You need to `#include "XBDateTime.h"` in `xbmc/music/Artist.h` for `CDateTime` to be declared.

**Comment by Razzeee** [comment]
Thanks, that solved it. Not sure how ```Album.h``` gets that dependency.

**Comment by Montellese** [comment]
Just add the same include there as well. Often headers get included through other headers but including them explicitly is preferred.

**Comment by Razzeee** [comment]
Minimized the code like you said, had to change some tests.
And I still have to figure out whats happening with the artist, doesn't work at all. Probably have missed something.

**Comment by Montellese** [comment]
Much cleaner thanks.

**Comment by Razzeee** [comment]
Is it still prefered to prefix the table/view columns with the datatype?
Their pretty mixed up right now. 80% have this the rest does not.

**Comment by Razzeee** [comment]
It's now fully working, but please review the last commit. I'm not sure if that is the 100% correct way to do this.

**Comment by MartijnKaijser** [comment]
lets see if it builds 
jenkins build this please

**Comment by Razzeee** [comment]
Tested it on mysql, everything seems fine.

Also rebased, that it is mergable

File: xbmc/music/GUIViewStateMusic.cpp [hunk 2]
```diff
@@ -188,6 +189,8 @@ CGUIViewStateMusicDatabase::CGUIViewStateMusicDatabase(const CFileItemList& item
       AddSortMethod(SortByArtist, sortAttribute, 557, LABEL_MASKS("%F", "", strAlbumLeft, strAlbumRight));  // Filename, empty | Userdefined, Userdefined
       // year
       AddSortMethod(SortByYear, 562, LABEL_MASKS("%F", "", strAlbumLeft, strAlbumRight));
+      // album date added
+      AddSortMethod(SortByDateAdded, sortAttribute, 570, LABEL_MASKS("%F", "", strAlbumLeft + " - " + strAlbumRight, "%a"));  // Filename, empty | Userdefined - Userdefined, dateAdded
```
Comment (Montellese) [inline xbmc/music/GUIViewStateMusic.cpp]: Is there a reason why this formatting differs from all the other sort methods?

File: xbmc/music/MusicDatabase.cpp [hunk 3]
```diff
@@ -3380,6 +3383,7 @@ bool CMusicDatabase::GetArtistsByWhere(const std::string& strBaseDir, const Filt
 
         pItem->GetMusicInfoTag()->SetDatabaseId(artist.idArtist, MediaTypeArtist);
         pItem->SetIconImage("DefaultArtist.png");
+        pItem->GetMusicInfoTag()->SetArtist(artist);
```
Comment (Montellese) [inline xbmc/music/MusicDatabase.cpp]: Is this necessary? Shouldn't this be covered by passing the `CArtist` object to `CFileItem`'s constructor?

File: xbmc/utils/test/TestDatabaseUtils.cpp [hunk 4]
```diff
@@ -177,13 +179,13 @@ TEST(TestDatabaseUtils, GetField_MediaTypeAlbum)
                                    DatabaseQueryPartSelect);
   EXPECT_STREQ(refstr.c_str(), varstr.c_str());
 
-  refstr = "albumview.idalbum";
+  refstr = "albumview.dateAdded";
   varstr = DatabaseUtils::GetField(FieldDateAdded, MediaTypeAlbum,
-                                   DatabaseQueryPartOrderBy);
+    DatabaseQueryPartSelect);
```
Comment (MartijnKaijser) [inline xbmc/utils/test/TestDatabaseUtils.cpp]: wrong indent?

**Comment by Montellese** [comment]
Spotted two more minors.

File: xbmc/music/MusicDatabase.cpp (hunk 3 above)
Comment (Razzeee) [inline xbmc/music/MusicDatabase.cpp]: Well it's not working without that line. As I wrote above if there is a better way to do it, please let me know.

File: xbmc/utils/test/TestDatabaseUtils.cpp (hunk 4 above)
Comment (Razzeee) [inline xbmc/utils/test/TestDatabaseUtils.cpp]: Not sure how that happened, will fix asap.

File: xbmc/music/GUIViewStateMusic.cpp (hunk 2 above)
Comment (Razzeee) [inline xbmc/music/GUIViewStateMusic.cpp]: Yes, that way we will see the date in the list. I actually think the year format should be changed in another PR.

Comment (Razzeee) [inline xbmc/music/GUIViewStateMusic.cpp]: Hrm, but this might also be a bad idea :/
http://kodi.wiki/view/advancedsettings.xml#musiclibrary
It's colliding a bit with that. We should have a way to show the value we're sorting by, but the user also wants to be able to format the strings by himself. And he probably thinks he is defining left | right. Not left - right.

Comment (Montellese) [inline xbmc/music/GUIViewStateMusic.cpp]: I would do it the same as the existing sort methods for now and then maybe make a PR to adjust them all and see what the feedback is.

Comment (Razzeee) [inline xbmc/music/GUIViewStateMusic.cpp]: Can do, but it feels kind of awkward, when you don't see the date added.
Especially when you go from:
Artist (you see the date added) -> Album (you don't see it) -> Song (you see the date added)

Comment (Montellese) [inline xbmc/music/GUIViewStateMusic.cpp]: TBH I have no clue why album listings are handled this special but I'd prefer to keep the current format with this addition and then discuss any visual improvements/changes in a separate PR because otherwise this whole addition is going to be blocked over such a minor point.

File: xbmc/music/MusicDatabase.cpp (hunk 3 above)
Comment (Montellese) [inline xbmc/music/MusicDatabase.cpp]: Ah it looks like `CFileItem(const CArtist& artist)` could need an improvement to call `CMusicInfoTag::SetArtist(const CArtist& artist)` instead of calling `CMusicInfoTag::SetArtist(const std::string& artist)` which only sets the name of the artist but not the rest of the details.

File: xbmc/music/GUIViewStateMusic.cpp (hunk 2 above)
Comment (Razzeee) [inline xbmc/music/GUIViewStateMusic.cpp]: Your absolutely right. Just wanted to mention that, will change.

Comment (Razzeee) [inline xbmc/music/GUIViewStateMusic.cpp]: Done

File: xbmc/music/MusicDatabase.cpp (hunk 3 above)
Comment (Razzeee) [inline xbmc/music/MusicDatabase.cpp]: Did that, working fine so far.

**Comment by Razzeee** [comment]
Will squash, when we find no more issues.

**Comment by Razzeee** [comment]
Squashed. Should be good to go.
Will do a new PR for the label format stuff soon. (See: https://github.com/xbmc/xbmc/pull/7739#discussion-diff-37146586)

**Comment by hudokkow** [comment]
jenkins build this please

**Comment by mkortstiege** [comment]
@Montellese, your button.
